faunadb-scala/src/main/scala/faunadb/FaunaClient.scala

```diff
@@ -190,11 +192,65 @@ class FaunaClient private (connection: Connection) {
+        case errorResponse => handleErrorResponse(errorResponse.statusCode(), errorResponse.body())
+      }
+      .recoverWith(handleNetworkExceptions)
+  }
+
+  /**
+    * Creates a subscription to the result of the given read-only expression. When
+    * executed, the expression must only perform reads and produce a single
+    * streamable type, such as a reference or a version. Expressions that attempt
+    * to perform writes or produce non-streamable types will result in an error.
+    * Otherwise, any expression can be used to initiate a stream, including
+    * user-defined function calls.
+    *
+    * @param expr the query to subscribe to.
+    * @param ec the `ExecutionContext` used to run the query asynchronously.
+    * @return A [[scala.concurrent.Future]] containing a [[java.util.concurrent.Flow.Publisher]] which yields element of
+    *         type [[faunadb.values.Value]]. The [[scala.concurrent.Future]] fails if the stream cannot be setup.
+    */
+  def stream(expr: Expr)(implicit ec: ExecutionContext): Future[Flow.Publisher[Value]] =
+    performStreamRequest(json.valueToTree(expr), None)
+
+  /**
+    * Similar to `stream` but it performs a GET and discards events prior to the document's snapshot.
+    * It sends the resulting snapshot as first event in the stream instead of a `start` event.
+    *
+    * @param expr the query to subscribe to.
+    * @param ec the `ExecutionContext` used to run the query asynchronously.
+    * @return A [[scala.concurrent.Future]] containing a [[java.util.concurrent.Flow.Publisher]] which yields element of
+    *         type [[faunadb.values.Value]]. The [[scala.concurrent.Future]] fails if the stream cannot be setup.
+    *
+    */
+  def streamDocument(expr: Expr)(implicit ec: ExecutionContext): Future[Flow.Publisher[Value]] = {
+    performStreamRequest(json.valueToTree(expr), None).map { valuePublisher =>
+      val documentValueFlowProcessor = new DocumentValueFlowProcessor(() => query(Get(expr)))
+      valuePublisher.subscribe(documentValueFlowProcessor)
+      documentValueFlowProcessor
+    }
+  }
+
+  private def performStreamRequest(body: JsonNode, timeout: Option[FiniteDuration])(implicit ec: ExecutionContext): Future[Flow.Publisher[Value]] = {
+    val javaTimeout = timeout.map(_.toJava).asJava
+    connection.performStreamRequest("POST", "stream", body, javaTimeout)
+      .toScala
+      .flatMap {
+        case successResponse if successResponse.statusCode() < 300 =>
+          // Subscribe a new FlowEventValueProcessor to consume the Body's Flow.Publisher
+          val flowEventValueProcessor = new BodyValueFlowProcessor()
+          successResponse.body().subscribe(flowEventValueProcessor)
+          Future.successful(flowEventValueProcessor)
+        case errorResponse =>
+          // The request failed, we need to consume the body manually for error reporting
+          connection.consumeBodyAsString(errorResponse)
```

Comment:
It feels a bit odd to send the response back to the `Connection` for parsing its body to a `String`. I think it was done this way, so that we could reuse it from the Java `FaunaClient` as well, right?

Mmh, maybe we can put that logic somewhere else outside the `Connection` (still in `common`)? Some kind of `StringProcessor` Future/helper?

(I'll try to think of a more concrete alternative myself too!)

Reply:
I agree, I was unsure about putting this helper in `Connection` too.
I will extract somewhere else 👍